### ISSUE TITLE
[Feature]: narrative test suite

### ISSUE THREAD

**Issue by srguiwiz** [microsoft/playwright#7756]
### Feature request

The purpose is testing of complete applications that have concepts like branch offices or clinics.  More explained below.

The solution would be instead of from a directory listing to get tests from a JavaScript generator function https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators .  One possible way to configure could be if `project.config.testDir` is a `function` instead of a `string`, but it could have its own name too.  The entry point to start implementing apparently would be `test/runner.ts` but it would affect `test/loader.ts`.  The functionality should be optional, only if such a function is provided.  If used, it would imply less parallelism, limiting to one worker per project.

There are three items of functionalitiy.  Currently we use very hacky workarounds.  To understand the need:

A branch office or clinic needs to be set up.  With proper authentication and verification, involving email, OAuth, etc., a staff memeber filling out information about themselves, etc. that process as **a test case can take a minute per user** (also for the clinic itself).  Outside IT systems are involved for the authentication and verification, which take any number of seconds to process, sometimes half a minute.  Not a problem, but it takes a while.

Those staff members need to exist and be used in further tests.  Not just one, but different roles, for example a doctor versus a technician, but also with different access rights, admin versus regular user.  Maybe a handful or a dozen users are needed, so that **further tests then can determine** that they cannot violate role specific limitations.  E.g. a non-admin user must not be able to change anyone else's records, or a tech may not be allowed to sign a diagnosis like a doctor.

Therefore, **executing tests in order is essential**:  A clinic must be created, professional users must be created, only then patients can be seen.

The creation of a clinic and of professional users itself are distinct test cases.  They are part of the system and as such deserved to be implemented as test cases.  It makes little sense to duplicate that ability some other way.  That reinforces the idea that executing tests in order is essential.

As a side note, it would be nice after creating a set of users, then to run some code that isn't test code proper, but that sets up "from now on use the users that have been created".  This does matter for the idea of execution in time:  If **the test running mechanism comes back to the generator function after each test**, that allows the generator function to have additional code in it, that "knows" after all users have been generated, to set something up for the following tests.  That step of setting up wouldn't work if the test running mechanism ahead of time asks for all tests to run.  (Workaround for this one requirement: Make and insert in order a fake test that does those intermediate setup steps.)

The next topic:  If you need ten users, maybe that would be ten times the same test case "register-a-new-user.test.ts".  The generator function could ten times give the same script to run.  That works better if the test running mechanism comes back with **next()** after each test and **the generator function gives it the same test file name to run again** next.  Each test run would be somewhat different with random or seed data.

There is more about randomization, but to keep it short, we can describe:  Running the same test case any number, three times, twenty time, a hundred times, is very valuable to what we have been doing.  Specifically there is a lengthy form, that takes half a minute or a minute to fill out, in part because of API calls to the server, not useless waiting, not lack of skill in test writing, just a lengthy form.  Imagine patient intake, health history, or imagine a complete tax form (which are longer for people with family or a business).  The point is:  A hundred or even ten or three of those, do not reasonably fit into a single test.  And, because of the many possible combinations of data (age, history, symptoms, etc.) randomization is a significant feature to actually achieve quality.

Therefore, the abilitiy to run the same test file multiple times, any number of repetions as seen fit that time (for example quick acceptance versus a broader sweep) is important.  (Difficult workaround:  Symlinks.  Not how it should be.  Parallelism, cleanup, naming, etc..)

It is important to do this within a test run, in order to have reporting.  We have so far done some of this with scripting outside loops, invoking Playwright for each repetition, but that means there isn't a singular final report.

Notably, a generator function can be invoked multiple times, also in parallel, e.g. if testing Chromium, Firefox, WebKit.

Related could be #7028 and maybe #7050 .

**Comment by pavelfeldman** [comment]
I read this as a request for the extensibility in the area of test generation. So that instead of discovering the tests ourselves, we would use third party function for that.

While we can do that, I think it'll lead to a number of problems we would need to handle:
- Setting up and tearing down fixtures per test. It is unclear how this fits your framework. Is test still an isolated web session that gets a fresh context or is that more of a single multi-step session?
- While you control the tests, what about describe / beforeEach / beforeAll - I don't think you want to control that externally

All it all, it looks outside the scope of what we envisioned for Playwright Test, but let me keep it open in case we get more interest from the community.

**Comment by srguiwiz** [comment]
@pavelfeldman Not new test generation, but different order of invocation, in order, and allowing repetition.  In case you haven't yet used JavaScript generator functions https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator please understand they are an improvement over iterators.  You might already know and use JavaScript generator functions, using **yield** to implement and providing an iterator with **next()**, I just want to make sure we don't have a misunderstanding.

This feature requests has nothing to do with test generation.

This feature request is meant to be conservative in requiring normal tests, as known.  We currently use fixtures and did not expect that to change.

The main difference is in "where does the list of tests come from"?

#7028 implemented a first step into the direction of this feature request.

Calling a generator function's iterator's `next()` instead of using `readDirAsync` to read `project.config.testDir` is a next step,

Also described, as an illustration, are specific use cases with branch offices or clinics, multiple users, and very long intake forms.

Again:  Normal tests.  More flexibility in invocation.  This feature request is about invocation, not generation :)

**Comment by srguiwiz** [comment]
A simple example in a `playwright.config.ts` could be:

``` typescript
const config: PlaywrightTestConfig = {
  projects: [
    {
      name: "launch-and-use-clinic",
      testSequence: function* testSequence() {
        yield "tests/launch-new-clinic.test.ts";
        for (let i = 0; i < 10; i++)
          yield "tests/add-staff-member.test.ts";
        for (let i = 0; i < 100; i++)
          yield "tests/fill-patient-intake-form.test.ts";
      },
    },
  ],
}
```

That would cause this project to launch 1 clinic, add 10 staff members (which is realistic, we take care of details), and add a 100 intake forms, which would give coverage because that test randomizes patient data, history, symtpoms, etc..

Each of those tests would be a normal test by itself.

We take care of any necessary configuration and communication.

`function* testSequence() {` … is a JavaScript generator function per [https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*) .

Some implementation detail would be whether `testSequence` precludes use of `testDir` in the same project etc..  Open to discussion here.

Should work at top level in `PlaywrightTestConfig` just like inside a project.

**Comment by pavelfeldman** [comment]
Generators are a very old tech, I'm surely familiar with them.

Your example doesn't use the generation aspect, it isn't dynamic, but I guess it is just a bad example. Imagining it being dynamic means that you would want to read the test status and based on that either run or not run other tests. This is against Playwright's test isolation, but let's for a moment imagine we want to support it.

Async/await are generators under the hood and they don't mix with yield well, so I assume the API you want would be more of an `const result = await run('test.spec.ts')` shape. It also needs to get the concept of parallelism into consideration. So it still makes me think that I understood your request the right way. I did not understand you want to control it on the file basis, it simplifies the things, but also defeats the purpose of fine-grained control of inter-dependent tests.

**Comment by srguiwiz** [comment]
The simple example shows two important features:  Sequential order and repetition.

The reason we use Playwright is because in many ways it is the best.  And, we try to use and fit in with existing patterns.

We believe end to end testing also should be able to cover "one user transaction", for example a patient intake on a lengthy web form, and in this case "the creation of a branch office or clinic, adding ten staff, and entering a hundred patient intakes".  As such it deserves a report, like any test suite.  The components are tests, fitting into the Playwright way of doing things.  The problem we are trying to address with this feature request is combination of tests into a "narration", "the first week of a new clinic", so that we can present a report on those tests passing.

Besides solving our problem, I also believe this proposal serves general needs, among those:

Rather long forms:  One cannot reasonably test part of such a long form.  The business chose to require a long patient intake form, and they are supposed to fill out all of it.  The impact being at least twofold:  One, a form easily takes half a minute, even if automated well.  Two, with twenty yes/no features (some actually a dozen, some hundreds of choices) that is 2²⁰ ≈ **one million combinations**, one cannot reasonably write specific test cases for a million different combinations within budget, on time, or maintainably, hence in addition to specific cases then the rest of that large possible space is desired to be reasonably **covered by randomized selections**.  Testing randomized combinations sometimes may reveal new failures, then deserving becoming a specific case.  The question how many randomized selection to run depends on priority.  It makes sense, and creates happiness, to claim "we ran a thousand randomize patient intakes", and here is the report.  While we can and have coded **repetition** outside Playwright, invoking it separately for each repetions, that prevents getting the benefits of a run of a test suite, having a single report to show.

**Dependency on sequential order to create a set of differently authorized users before running other tests**:  Setting up a clinic, and adding staff memebers are expensive in terms of time use.  As mentioned, other services are involved, OAuth, email.  A minute each is normal.  Faking it wouldn't be full end to end testing.  One cannot expect a test to actually create two staff members before testing for example whether privileges are observed correctly.  The creation of staff members already exists as a distinct test.  Because of **how long it takes**, and because of the **DRY** (don't repeat yourself) principle, it makes sense to run those tests first, then using those staff members in further testing.  That, I must believe, is or will become of general interest.  Again, should be all in one test suite, for a single report.

Therefore, while intentionally short, that example uses the generation aspect, showing **sequencing** and **repetition**, with the `for` loops.

I have so far in that simple example omitted the aspect of **doing something between tests**.  One goal (with the second of the two possible implementation steps) is to allow any code to run at the apparently correct time in the generator function, for example grabbing something from the newly created clinic and changing a configuration to then run all following tests with that clinic.  The simple example doesn't show that.  One reason is to keep it simpler to implement, step one.  The other reason is, if only implementing step one, inconvenient workarounds are possible meanwhile.

In the first implementation step that code of doing something between tests wouldn't run at the right time, so I didn't want to distract with that.  Using a generator function allows to proceed to the second step in a general way.

The title of "narrartive test suite" expresses the idea "this is a narration of what happens, this, then that, then that", with repetition, for randomized testing of a space too large to completely cover with specific tests.

Playwright is awesome in many ways, an excellent foundation for automated testing.  Hopefully this will make it easier to test a complete business within a single test suite, with a single report resulting from it.

**Comment by srguiwiz** [comment]
Regarding parallelism, if one choses `testSequence` that should or would have to imply limitation to one worker for that project.

**Comment by srguiwiz** [comment]
There shouldn't be a problem with yield.  Per https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/yield

> Once paused on a yield expression, the generator's code execution remains paused until the generator's next() method is called.

A rarely demonstrated feature in example code, something I have not used yet:

> If an optional value is passed to the generator's next() method, that value becomes the value returned by the generator's current yield operation.

Getting back something could be interesting or nice.  I had not thought of or expected that.

On the implementation side, in Playwright's `test/runner.ts` I would assume something like you mentioned `const result = await run('test.spec.ts')` to be called before calling `next()`, or `next(result)`.

Sidestepping that, an easier implementation, once called "a first step" above, is to use the generator function to get a complete list of files synchronously like `readdirSync`.  It would require a change at least where `loadTestFile` currently only allows a file to be loaded once.  It would be a progress over the current situation, by allowing various ways to author sequential order, without requiring alphabetic order, and by allowing repetition.  It is a compromise.  You have deeper insight into your code and may prefer skip this compromise to proceed to a fully asynchronous implemention.

After a test fails, it would be up to subsequent tests, for example if consequentially configuration misses required information (not knowing what clinic to use, or what staff members), then for a subsequent test to fail with "missing necessary piece of configuration".  Tests should fail anyway if a piece of configuration is missing.  That is easy to understand, for everyone on the team:  If one didn't succeed at setting up a clinic and adding staff members, then one cannot start seeing patients.  The purpose of a project using `testSequence` is a complete narration, made of the same pieces as isolated tests.  Failing to set up a clinic means a complete failure of that narration.  It is understandable in part because it reflects real life.

One goal is automated testing to be maintainable by subsequent developers.  Will source code of automated testing be readable?  Can people figure where to put additional tests?

A generator function allows different styles or authoring to develop in the community, maybe similar to how `playwright.config.ts` allows any style of coding to make the required `PlaywrightTestConfig` object.

**Comment by srguiwiz** [comment]
We sometimes customize invocations of the same test file with environment variables.  Here would be an example to do that with `testSequence`:


``` typescript
const config: PlaywrightTestConfig = {
  projects: [
    {
      name: "launch-and-use-clinic",
      testSequence: function* testSequence() {
        yield "tests/launch-new-clinic.test.ts";
        yield {
          test: "tests/add-staff-member.test.ts",
          env: {
            customEnvVar: JSON.stringify({
              "role": "dr",
              "permissions": "admin",
            })
          },
        };
        yield {
          test: "tests/add-staff-member.test.ts",
          env: {
            customEnvVar: JSON.stringify({
              "role": "tech",
              "permissions": "staff",
            })
          },
        };
        yield {
          test: "tests/add-staff-member.test.ts",
          env: {
            customEnvVar: JSON.stringify({
              "role": "billing",
              "permissions": "staff",
            })
          },
        };
        for (let i = 0; i < 7; i++)
          yield "tests/add-staff-member.test.ts";
        for (let i = 0; i < 100; i++)
          yield "tests/fill-patient-intake-form.test.ts";
      },
    },
  ],
}
```

For an implementation of `testSequence` that means "support some way of getting parameters to a test".  Here shown is using environment variables.  You can treat the environment variables as an object whose values are strings, ignore how they are made in the example via `JSON.stringify`.

At a higher level, in this example the generic add-staff-member test is told not to be completely random, but with the first three invocations to create specific kind of staff members, by presetting some of their many properties.  The remaining test runs are all going to generate all data as random as they are coded to be.

**Comment by pavelfeldman** [comment]
Your request is clear. Testing frameworks use term "test generation" in a sense of generating sequences of parametrized tests. So I still I think I got your request exactly right. What I don't understand is what is missing. Today you can run:

```js
for (let i = 0; i < 10; ++i) {
  test('foo'+i, async ({}) => {})
}
```

will generate 100 tests, so you can replace calls to yield in your snippet with calls to test pretty much.

**Comment by srguiwiz** [comment]
Excellent, that existing functionality fulfills the need.

One official confirmation would create certainty going forward.

It wasn't clear so far that those `test` would usually or always execute in order, from what I found documented.  Maybe I overlooked that, or it wasn't said explicitly.  Having seen old documentation where `beforeEach` and `afterEach` could be in any order with several `test` inside a `whatever.test.ts` it didn't seem clear that sequential order necessarily was being maintained.

If in `whatever.test.ts` it reads

``` typescript
import { test } from "@playwright/test";
for (let i = 0; i < 500; ++i) {
  console.log('defining', i);
  test('foo'+i, async ({}) => {
    console.log('running', i);
  })
}
```

and running configured to one worker and chromium only, then output says

```
defining 0
defining 1
…
defining 498
defining 499

Running 500 tests using 1 worker

defining 0
defining 1
…
defining 498
defining 499
  ✓ whatever.test.ts:4:3 › [chromium] foo0 (9ms)
running 0
  ✓ whatever.test.ts:4:3 › [chromium] foo1 (4ms)
running 1
…
  ✓ whatever.test.ts:4:3 › [chromium] foo498 (1ms)
running 498
  ✓ whatever.test.ts:4:3 › [chromium] foo499 (1ms)
running 499

  500 passed (2s)
```

That seems to be indicative, which makes sense, that `test` line of source code defines a test, and some object gets created, all such objects get collected, and then one after the other the actual test code gets invoked.

Observation seems to indicate this happens in order, in one worker.

Notably, if there is a failure in a test, we observe another batch of
```
defining 0
defining 1
…
defining 498
defining 499
```
before it continues with the next test.  That also makes sense, it is another run of the content of `whatever.test.ts` in a new worker instance.  That is consistent with documentation.

What would be helpful is official confirmation that multiple `test('foo'+i, async({})=>{})` calls in a single `whatever.test.ts` by design will run in order, despite other parallelism.  It might be deducted from reading source code, but you know that code.

That would set it apart from alphabetic ordering of `*.test.ts` files by name if running with a single worker, which is different but also useful predictable behavior.

Thank you for having followed the thread and having found a solution.

**Comment by pavelfeldman** [comment]
#### Why was this issue closed?

We are prioritizing the features based on the upvotes, recency and activity in the issue thread. It looks like this issue only has a handful of upvotes, has not been touched recently and/or we lack sufficient feedback to act on it. We are closing issues like this one to keep our bug database maintainable. Please feel free to open a new issue and link this one to it if you think this is a mistake.
